pymchelper/readers/shieldhit/general.py

```diff
@@ -122,15 +123,26 @@ def get_reader(self):
 
         # TODO add ZIP file unpacking
 
+        # Trying to decompress bdz-file. If it works, saves a temporary bdo-file which will be used for reading later
+        # zip_name is the addition to the file name for the new bdo-file, compared to the original: nothing in case
+        # original is already bdo, and "_TEMP.bdo" in case original was a bdz.
+
+        with open(self.filename, "rb") as f:
+            try:
+                bdo = zlib.decompress(f.read())
+            except zlib.error:
+                bdo = f.read()
+
         # magic number was introduced together with first token-based BDO file format (BDO2016)
         # presence of magic number means we could have BDO2016 or BDO2019 format
-        if file_has_sh_magic_number(self.filename):
+        if file_has_sh_magic_number(bdo):
             reader = SHReaderBDO2019
 
             # format tag specifying binary standard was introduced in SH12A v0.7.4-dev on  07.06.2019 (commit 6eddf98)
             file_format = read_token(self.filename, SHBDOTagID.format)
```

Comment:
you need to change the read_token() function to accept the bdo object instead of self.filename...

and basically you have to do this with every function which used to touch the original file. :)

Reply:
Yes, that is (one of the few parts :), which is) clear to me. But I thought this would have other effects further down in the code where it expects the filename, and not the object, for example in the np.fromfile method. But I´ll try some more later, maybe it is more easy than I think